### PR TITLE
completely shutdown dyn_params_handler_ (s) 

### PR DESCRIPTION
**Pull request by GoesM**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #4496  |
| Primary OS tested on | (Ubuntu, MacOS, Windows) |
| Robotic platform tested on | (Steve's Robot, gazebo simulation of Tally, hardware turtlebot) |
| Does this PR contain AI generated software? | (No; Yes and it is marked inline in the code) |

---

## Description of contribution in a few bullet points

result from experiments 
In `amcl` node, the following code could completely shutdown the callback threads properly:
```cpp
  // shutdown and reset dynamic parameter handler
  remove_on_set_parameters_callback(dyn_params_handler_.get());
  dyn_params_handler_.reset();
```

## Future work that may be required in bullet points
Here's a todo list for my future work, I think I'd check out package by package: 
- [x] nav2_amcl (fixed)
- [x] nav2_behavior_tree (no dynamic_param)
- [x] nav2_behaviors (no dynamic_param)
- [x] nav2_bringup (no dynamic_param)
- [x] nav2_bt_navigator
- [x] nav2_collision_monitor
- [x] nav2_common
- [x] nav2_constrained_smoother
- [x] nav2_controller
- [x] nav2_core
- [x] nav2_costmap_2d
- [x] nav2_docking
- [x] nav2_dwb_controller
- [x] nav2_graceful_controller
- [x] nav2_lifecycle_manager
- [x] nav2_map_server
- [x] nav2_mppi_controller
- [x] nav2_msgs (no dynamic param)
- [x] nav2_navfn_planner
- [x] nav2_planner
- [x] nav2_regulated_pure_pursuit_controller
- [x] nav2_rotation_shim_controller
- [x] nav2_rviz_plugins
- [x] nav2_simple_commander
- [x] nav2_smac_planner
- [x] nav2_smoother
- [x] nav2_system_tests 
- [x] nav2_theta_star_planner
- [x] nav2_util
- [x] nav2_velocity_smoother
- [x] nav2_voxel_grid
- [x] nav2_waypoint_follower

This PR is opened as a demo. If there are no issues here, I will quickly check and correct all remaining packages, and then update this PR in one go.

### CODE REVIEW

**Review by SteveMacenski — APPROVED** [review]
LGTM, ping me back when all done and I can merge 🙂

**Comment by GoesM** [comment]
There'are still changes for plugins, I would add soon.

**Comment by SteveMacenski** [comment]
This seems to cause a number of crashes, according to the test results

**Comment by GoesM** [comment]
It seems that in the test, some dyn_param_handler instances might not be registered, which can lead to related errors during resource release. I am currently making adjustments.

**Comment by GoesM** [comment]
The latest CI test failed in `nav2_behavior_tree` , which seems not related to my change (I think) ?

**Comment by SteveMacenski** [comment]
Lets try that once more then - retriggering. Can you also rebase? Maybe there's something weird going on with caching if this branch has an old base.

**Comment by GoesM** [comment]
opps ....  So sorry for that many weird things happened after rebase, hoping it's the final change ....

**Comment by GoesM** [comment]
still failed in `nav2_behavior_tree` , 

I reviewed the dependencies of nav2_behavior_tree, and it appears that it does not depend on any of the packages I modified.
SO I feel this failure is likely unrelated to my changes. 

Perhaps, to be more cautious, I could also submit new PR for additional try?

File: nav2_costmap_2d/plugins/inflation_layer.cpp [hunk 1]
```diff
@@ -78,7 +78,12 @@ InflationLayer::InflationLayer()
 
 InflationLayer::~InflationLayer()
 {
+  auto node = node_.lock();
+  if (dyn_params_handler_ && node) {
+    node->remove_on_set_parameters_callback(dyn_params_handler_.get());
+  }
   dyn_params_handler_.reset();
+  node.reset();
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/inflation_layer.cpp:86]: I don't think this is necessary? (or all)

File: nav2_regulated_pure_pursuit_controller/include/nav2_regulated_pure_pursuit_controller/parameter_handler.hpp [hunk 2]
```diff
@@ -81,13 +81,14 @@ class ParameterHandler
...
    */
-  ~ParameterHandler() = default;
+  ~ParameterHandler();
 
   std::mutex & getMutex() {return mutex_;}
 
   Parameters * getParams() {return &params_;}
 
 protected:
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node_;
```
Comment (SteveMacenski) [inline nav2_regulated_pure_pursuit_controller/include/nav2_regulated_pure_pursuit_controller/parameter_handler.hpp:91]: This must be a weak pointer if stored

File: nav2_graceful_controller/include/nav2_graceful_controller/parameter_handler.hpp [hunk 3]
```diff
@@ -70,13 +70,15 @@ class ParameterHandler
...
    */
-  ~ParameterHandler() = default;
+  ~ParameterHandler();
 
   std::mutex & getMutex() {return mutex_;}
 
   Parameters * getParams() {return &params_;}
 
 protected:
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node_;
```
Comment (SteveMacenski) [inline nav2_graceful_controller/include/nav2_graceful_controller/parameter_handler.hpp:80]: This must be a weak pointer if stored

File: nav2_dwb_controller/dwb_plugins/include/dwb_plugins/kinematic_parameters.hpp [hunk 4]
```diff
@@ -150,6 +150,7 @@ class KinematicsHandler
   using Ptr = std::shared_ptr<KinematicsHandler>;
 
 protected:
+  nav2_util::LifecycleNode::SharedPtr node_;
```
Comment (SteveMacenski) [inline nav2_dwb_controller/dwb_plugins/include/dwb_plugins/kinematic_parameters.hpp:153]: This must be a weak pointer if stored

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

**Comment by SteveMacenski** [comment]
I triggered one more time... but a few issues to resolve here

**Comment by SteveMacenski** [comment]
> I notice that it still failed in nav2_behavior_tree. 😵

I just checked our nightlies, its failing there too for some reason. You can ignore that here, not your fault! 

> Shall I create another fork and submit a new PR ? or I still push my lateset changes here ?

You can just update this PR :-)

File: nav2_costmap_2d/plugins/inflation_layer.cpp (hunk 1 above)
Comment (GoesM) [reply nav2_costmap_2d/plugins/inflation_layer.cpp]: I've removed them now.

File: nav2_regulated_pure_pursuit_controller/include/nav2_regulated_pure_pursuit_controller/parameter_handler.hpp (hunk 2 above)
Comment (GoesM) [reply nav2_regulated_pure_pursuit_controller/include/nav2_regulated_pure_pursuit_controller/parameter_handler.hpp]: sure. I've changed them now.

File: nav2_graceful_controller/include/nav2_graceful_controller/parameter_handler.hpp (hunk 3 above)
Comment (GoesM) [reply nav2_graceful_controller/include/nav2_graceful_controller/parameter_handler.hpp]: sure. I've changed them now.

File: nav2_dwb_controller/dwb_plugins/include/dwb_plugins/kinematic_parameters.hpp (hunk 4 above)
Comment (GoesM) [reply nav2_dwb_controller/dwb_plugins/include/dwb_plugins/kinematic_parameters.hpp]: sure. I've changed them now.

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
Great, thanks!! Merged :-) This was a good find!

**Pull request merged**